src/serlio/prtMaterial/ArnoldMaterialNode.cpp

```diff
@@ -0,0 +1,575 @@
+	MCHECK(status);
+	if (!inMetadata) {
+		return MStatus::kSuccess;
+	}
+
+	// TODO: this should really be a traversal over the (potentially cyclic!) dependency
+	// graph to find all the relevant directly and indirectly connected mesh nodes
+	// for now, just follow direct or indirect connections via group parts nodes
+	MString meshName;
+	bool meshFound = false;
+	bool searchEnded = false;
+
+	for (MPlug curPlug = plug; !searchEnded;) {
+		searchEnded = true;
+		MPlugArray connectedPlugs;
+		curPlug.connectedTo(connectedPlugs, false, true, &status);
+		MCHECK(status);
+		if (!connectedPlugs.length()) {
+			return MStatus::kFailure;
+		}
+		for (const auto& connectedPlug : makePlugArrayConstWrapper(connectedPlugs)) {
+			MFnDependencyNode connectedDepNode(connectedPlug.node(), &status);
+			MCHECK(status);
+			MObject connectedDepNodeObj = connectedDepNode.object(&status);
+			MCHECK(status);
+			if (connectedDepNodeObj.hasFn(MFn::kMesh)) {
+				meshName = connectedDepNode.name(&status);
+				MCHECK(status);
+				meshFound = true;
+				break;
+			}
+			if (connectedDepNodeObj.hasFn(MFn::kGroupParts)) {
+				curPlug = connectedDepNode.findPlug("outputGeometry", true, &status);
+				MCHECK(status);
+				searchEnded = false;
+				break;
+			}
+		}
+	}
+
+	if (!meshFound) {
+		return MStatus::kSuccess;
+	}
+
+	adsk::Data::Associations inAssociations(inMetadata);
+	adsk::Data::Channel* inMatChannel = inAssociations.findChannel(gPRTMatChannel);
+
+	if (!inMatChannel) {
+		return MStatus::kSuccess;
+	}
+
+	adsk::Data::Stream* inMatStream = inMatChannel->findDataStream(gPRTMatStream);
+	if (!inMatStream) {
+		return MStatus::kSuccess;
+	}
+
+	//find all existing prt materials
+	const adsk::Data::Structure* fStructure = adsk::Data::Structure::structureByName(gPRTMatStructure.c_str());
+	if (!fStructure) {
+		return MStatus::kFailure;
+	}
+
+	std::set<std::wstring> shaderNames;
+	std::list<std::pair<const MObject, const MaterialInfo>> existingMaterialInfos;
+	MItDependencyNodes itHwShaders(MFn::kPluginHardwareShader, &status);
+	MCHECK(status);
+	for (const auto& hwShaderNode : MItDependencyNodesWrapper(itHwShaders)) {
+		MFnDependencyNode fnDependencyNode(hwShaderNode, &status);
+		MCHECK(status);
+		MString shaderNodeName = fnDependencyNode.name(&status);
+		MCHECK(status);
+		shaderNames.insert(shaderNodeName.asWChar());
+
+		const adsk::Data::Associations* materialMetadata = fnDependencyNode.metadata(&status);
+		MCHECK(status);
+
+		if (!materialMetadata) {
+			continue;
+		}
+
+		adsk::Data::Associations materialAssociations(materialMetadata);
+		adsk::Data::Channel* matChannel = materialAssociations.findChannel(gPRTMatChannel);
+		if (!matChannel) {
+			continue;
+		}
+
+		adsk::Data::Stream* matStream = matChannel->findDataStream(gPRTMatStream);
+		if (!matStream || matStream->elementCount() != 1) {
+			continue;
+		}
+
+		adsk::Data::Handle matSHandle = matStream->element(0);
+		if (!fStructure || !matSHandle.usesStructure(*fStructure)) {

```

Comment:
checking fStructure again is redundant, you already return above in that case

Reply:
good catch, that was a left-over from some refactoring